Applications/Spire/Source/OrderImbalanceIndicator/LocalOrderImbalanceIndicatorModel.cpp

```diff
@@ -0,0 +1,39 @@
+#include "Spire/OrderImbalanceIndicator/LocalOrderImbalanceIndicatorModel.hpp"
+#include "Spire/Spire/QtPromise.hpp"
+
+using namespace Nexus;
+using namespace Spire;
+
+std::tuple<boost::signals2::connection,
+    QtPromise<std::vector<Nexus::OrderImbalance>>>
+    LocalOrderImbalanceIndicatorModel::subscribe(
+    const boost::posix_time::ptime& start,
+    const boost::posix_time::ptime& end,
+    const OrderImbalanceSignal::slot_type& slot) {
+  m_signals.emplace_back(OrderImbalanceSignalConnection{
+      OrderImbalanceSignal(), start, end});
+  return std::tuple<boost::signals2::connection,
+    QtPromise<std::vector<OrderImbalance>>>(
+      m_signals.back().m_imbalance_signal.connect(slot),
+      QtPromise([start, end, this] {
+        auto requested_imbalances = std::vector<OrderImbalance>();
+        for(auto& imbalance : m_imbalances) {
+          if(imbalance.m_timestamp >= start &&
+              imbalance.m_timestamp <= end) {
```

Comment:
Construct the result outside of the promise rather than from inside of the promise to avoid potential race conditions.

Reply:
Would this involve a call to a function that returns the data?

Reply:
No, it's more along the lines of this:

    auto imbalances = std::vector<OrderImbalance>();
    for(auto& imbalance : m_imbalances) {
      if(imbalance.m_timestamp >= start &&
          imbalance.m_timestamp <= end) {
        imbalances.push_back(imbalance);
      }
    }
    return QtPromise([imbalances = std::move(imbalances)] mutable {
      return std::move(imbalances);
    });

You build the result outside of the promise.

Reply:
Whoops, I completely misinterpreted that code; I thought it was a then(...) statement.

Reply:
Why are the imbalances moved on returning from the promise?